README.md

```diff
@@ -999,6 +1004,47 @@ Specifying these flags from the command-line can be cumbersome and you may forge
 
 **TODO: Discuss desired/not desired options with MH.**
 
+*** Puneet: Surely we want Release as the default setting for BUILD_TYPE? ***
+
+*** Puneet: How do I specify mpi (or any other non-standard) compilers? *** 
+
+*** Puneet: Can we list a preset file that does the equivalent of the following: 
+```bash
+--enable-suppress-pdf-doc
+--enable-MPI
+--enable-multiple_teuchos_libraries
+--enable-symbolic-links-for-headers
+--with-blacs=/home/mheil/local/mumps_and_scalapack_default_installation/lib
+--with-scalapack=/home/mheil/local/mumps_and_scalapack_default_installation/lib/libscalapack.a
+--with-pord=/home/mheil/local/mumps_and_scalapack_default_installation/lib/libpord.a 
+--with-mumps=/home/mheil/local/mumps_and_scalapack_default_installation
+--with-blas=/home/mheil/local/lib/blas/blas.a
+--with-lapack=/home/mheil/local/lib/lapack/lapack.a
+--with-hypre=/home/mheil/local/hypre_default_installation_mpi
+--with-trilinos=/home/mheil/local/trilinos_default_installation_mpi
+--with-mpi-self-tests="mpirun -np 2"
+--with-mpi-self-tests-variablenp="mpirun -np OOMPHNP"
+--with-gmsh-self-tests="/home/mheil/gmesh/gmsh-3.0.6-source/build/gmsh"
+--with-mpi-include-directory=/usr/lib/openmpi/include
+--with-boost=/home/mheil/local/boost_default_installation
+--with-gmp=/home/mheil/local/gmp_default_installation
+--with-mpfr=/home/mheil/local/mpfr_default_installation
+--with-cgal=/home/mheil/local/cgal_default_installation
+CXXFLAGS="-O3 -Wall -std=c++11"
+CFLAGS="-O3"
+FFLAGS="-O3"
+FFLAGS_NO_OPT="-O0"
+CXX=mpic++
+CC=mpicc
+F77=mpif77
+FC=mpif90
+LD=mpif77
+```
```

Comment:
See below for a breakdown. A lot of the changes required to specify your own installation have just been made today so there may still be some bugs in there... There's also a circular dependency issue with Hypre and Trilinos that stops us from linking to our own build of BLAS/LAPACK that I've only just bumped into today (because I've just been using a previously installed version of OpenBLAS for so long). It's on the to-do list to fix!

Might be worth having a catch-up call to discuss this stuff when you're available

#### `--enable-suppress-pdf-doc`

> I haven't sorted out the documentation yet but I will make a flag named `OOMPH_ENABLE_SUPPRESS_PDF_DOC[=ON/OFF]` that the user can supply

#### `--enable-MPI`

> `OOMPH_ENABLE_MPI=ON/OFF`

#### `--enable-symbolic-links-for-headers`

> `OOMPH_ENABLE_SYMBOLIC_LINKS_FOR_HEADERS=ON/OFF`

#### `--with-cgal=/home/mheil/local/cgal_default_installation`/`--with-boost=/home/mheil/local/boost_default_installation`/`--with-mpfr=/home/mheil/local/mpfr_default_installation`/`--with-gmp=/home/mheil/local/gmp_default_installation`/`--with-blas=/home/mheil/local/lib/blas/blas.a`/`--with-lapack=/home/mheil/local/lib/lapack/lapack.a`/`--with-hypre=/home/mheil/local/hypre_default_installation_mpi`/`--with-trilinos=/home/mheil/local/trilinos_default_installation_mpi`

> Add `"OOMPH_USE_<LIB>_FROM": "/path/to/lib"` to your presets file or pass `-DOOMPH_USE_<LIB>_FROM=/path/to/lib` at configure time, where `<LIB>` is the library in capital letters, e.g. `-DOOMPH_USE_CGAL_FROM=/home/joe_user/cgal-installation`.
> **Example:**
>
> ```bash
> cmake -DOOMPH_USE_BLAS_FROM=/home/mheil/local/ \
>   -DOOMPH_USE_LAPACK_FROM=/home/mheil/local/ \
>   -DOOMPH_USE_GMP_FROM=/home/mheil/local/gmp_default_installation \
>   -DOOMPH_USE_MPFR_FROM=/home/mheil/local/mpfr_default_installation \
>   -DOOMPH_USE_BOOST_FROM=/home/mheil/local/boost_default_installation \
>   -DOOMPH_USE_CGAL_FROM=/home/mheil/local/cgal_default_installation
>   -DOOMPH_USE_HYPRE_FROM=/home/mheil/local/hypre_default_installation_mpi \
>   -DOOMPH_USE_TRILINOS_FROM=/home/mheil/local/trilinos_default_installation_mpi \
> ```
>
> When we start using preset files, you can just store something like this:
>
> ```json
> {
>   "version": 5,
>   "configurePresets": [
>     {
>       "inherits": "base",
>       "name": "cgal_preset",
>       "displayName": "cgal_preset",
>       "cacheVariables": {
>         "OOMPH_USE_BLAS_FROM": "/home/mheil/local/",
>         "OOMPH_USE_LAPACK_FROM": "/home/mheil/local/",
>         "OOMPH_USE_GMP_FROM": "/home/mheil/local/gmp_default_installation",
>         "OOMPH_USE_MPFR_FROM": "/home/mheil/local/mpfr_default_installation",
>         "OOMPH_USE_BOOST_FROM": "/home/mheil/local/boost_default_installation",
>         "OOMPH_USE_CGAL_FROM": "/home/mheil/local/cgal_default_installation",
>         "OOMPH_USE_HYPRE_FROM": "/home/mheil/local/hypre_default_installation_mpi",
>         "OOMPH_USE_TRILINOS_FROM": "/home/mheil/local/trilinos_default_installation_mpi"
>       }
>     },
>     // ...
>   ]
>   // ...
> }
> ```
>
> **NOTE:** It's important that the third-party libraries installed by the user (apart from GMP and MPFR) use the CMake build system otherwise CMake won't be able to recognise the installed libraries. The versions of Hypre, Trilinos, Boost and CGAL that I'm currently using are sufficient. (Well, there's a bug in CGAL that means we have to download it by hand and patch it ourselves for now.)
>
> **IMPORTANT:** Don't install and link to your own version of CGAL for now due to the issue mentioned [here](https://github.com/CGAL/cgal/issues/7121). When they release a version with the patch, we can allow users to link to it with the `OOMPH_USE_CGAL_FROM` configure option

#### `--enable-multiple_teuchos_libraries`

> Internal variable to handle additional Trilinos libraries; no longer needed.

#### `--with-blacs=/home/mheil/local/mumps_and_scalapack_default_installation/lib`/`--with-pord=/home/mheil/local/mumps_and_scalapack_default_installation/lib/libpord.a`/`--with-scalapack=/home/mheil/local/mumps_and_scalapack_default_installation/lib/libscalapack.a`/`--with-mumps=/home/mheil/local/mumps_and_scalapack_default_installation`

> These four are going to be a pain; the old versions of these libraries didn't use CMake so they're nasty to integrate into CMake libraries but the one we're using from GitHub works nicely.

#### `--with-mpi-self-tests="mpirun -np 2"`

> To make the MPI execution command platform-independent, it is no longer a good idea to hardcode `mpirun -np`. Instead we construct it from variables set by `FindMPI.cmake` (a CMake helper script that is read to find MPI "stuff"). When you run `ctest` on an MPI-enabled demo driver, it will run with `${MPIEXEC_EXECUTABLE} ${MPIEXEC_NUMPROC_FLAG} <N>` which will be substituted with the values appropriate for your build/system (e.g. `mpirun -np <N>` or `mpiexec -n <N>`).
>
> All YOU need to specify is the number of processes you want to use with the (very new) `OOMPH_MPI_NUM_PROC` flag.

#### `--with-mpi-self-tests-variablenp="mpirun -np OOMPHNP"`

> Similar to `--with-mpi-self-tests="mpirun -np 2"`. However, is it a good idea to let people change this one? (Since some of the demo drivers rely on `OOMPHNP` being there so they can `sed` replace it for a number)

#### `--with-gmsh-self-tests="/home/mheil/gmesh/gmsh-3.0.6-source/build/gmsh"`

> Now on the to-do list! I haven't added functionality for storing the path to `gmsh` yet. Glad you pointed this one out...

#### `--with-mpi-include-directory=/usr/lib/openmpi/include`

> `OOMPH_USE_MPI_FROM=/usr/lib/openmpi/`

#### `CXXFLAGS="-O3 -Wall -std=c++11"`/`CFLAGS="-O3"`/`FFLAGS="-O3"`

> Debug/optimisation flags handled with `CMAKE_BUILD_TYPE="Debug"` or `CMAKE_BUILD_TYPE="Release"`. If you REALLY have to specify a flag, use
>
> ```cmake
> add_compile_options(-Wall -Wextra -pedantic -Werror)
> add_compile_definitions(DUMMY_DEFINE) # -DDUMMY_DEFINE
> ```
>
> Target-based stuff is done like so:
>
> ```cmake
> target_compile_options(<my-target> PRIVATE -Wall -Wextra -pedantic -Werror)
> target_compile_definitions(<my-target> PRIVATE DUMMY_DEFINE)
> ```
>
> Here, `PRIVATE` indicates that you only want `<my-target>` to get the `DUMMY_DEFINE` definition and not anything that links to it. You can replace it with `PUBLIC` to indicate that anything linking to it should also get that definition.

#### `FFLAGS_NO_OPT="-O0"`

> Hmm, where is this necessary?
>
> To compile specific targets without optimisation, you can use the code below:
>
> ```cmake
> set_source_files_properties(predicates.cxx PROPERTIES COMPILE_FLAGS -O0)
> ```

#### `CXX=mpic++`/`CC=mpicc`/`F77=mpif77`/`FC=mpif90`/`LD=mpif77`

> Set `MPI_C_COMPILER`, `MPI_CXX_COMPILER`, `MPI_Fortran_COMPILER`; do NOT set `CC`, `CXX` or `F77`.
>
> **Example:**
>
> ```json
> {
>   "version": 5,
>   "configurePresets": [
>     {
>       "inherits": "base",
>       "name": "mypreset",
>       "displayName": "mypreset",
>       "generator": "Ninja",
>       "binaryDir": "${sourceDir}/build",
>       "cacheVariables": {
>         "MPI_C_COMPILER": "mpicc",
>         "MPI_CXX_COMPILER": "mpic++",
>         "MPI_Fortran_COMPILER": "mpif77"
>       }
>     },
>     // ...
>   ]
>   // ...
> }
> ```